File: packages/e2e-test/src/__tests__/completion.test.ts
/**
 * End-to-End Tests for completion generation.
 *
 * These tests verify that the CLI correctly:
 * - Generates shell completions after tool installation
 * - Creates completion files for both zsh and bash
 * - Includes proper completion functions and commands
 */
import { beforeAll, describe, expect, it } from 'bun:test';
// oxlint-disable-next-line import/no-unassigned-import
import '@dotfiles/testing-helpers';
import { Architecture, Platform } from '@dotfiles/core';
import { withMockServer } from './helpers/mock-server';
import { TestHarness } from './helpers/TestHarness';

describe('E2E: completion generation', () => {
  withMockServer((b) =>
    b.withScript(
      '/mock-install-for-cmd-completion-test.sh',
      'tools/curl-script--cmd-completion-test/mock-install-for-cmd-completion-test.sh',
    )
  );

  const platformConfigs: ReadonlyArray<{
    platform: Platform;
    architecture: Architecture;
    name: string;
  }> = [
    { platform: Platform.MacOS, architecture: Architecture.Arm64, name: 'macOS ARM64' },
    { platform: Platform.Linux, architecture: Architecture.X86_64, name: 'Linux x86_64' },
  ];

  for (const config of platformConfigs) {
    describe(`${config.name}`, () => {
      const harness: TestHarness = new TestHarness({
        testDir: import.meta.dir,
        configPath: 'fixtures/main/config.yaml',
        platform: config.platform,
        architecture: config.architecture,
      });

      beforeAll(async () => {
        await harness.clean();
        const generateResult = await harness.generate();
        expect(generateResult.code).toBe(0);
      });

      describe('completion generation', () => {
        it('should generate completions after tool installation', async () => {
          // Verify shim was created
          await harness.verifyShim('curl-script--cmd-completion-test');

          // Install the tool explicitly
          const installResult = await harness.install(['curl-script--cmd-completion-test']);
          expect(installResult.code).toBe(0);

          // Verify completion files were generated
          const zshCompletionPath = harness.getCompletionPath('curl-script--cmd-completion-test', 'zsh');
          const bashCompletionPath = harness.getCompletionPath('curl-script--cmd-completion-test', 'bash');

          await harness.verifyFile(zshCompletionPath);
          await harness.verifyFile(bashCompletionPath);

          // Verify completion content contains expected strings
          const zshContent = await harness.readFile(zshCompletionPath);
          expect(zshContent).toContain('#compdef curl-script--cmd-completion-test');
          expect(zshContent).toContain('_curl-script--cmd-completion-test');

          const bashContent = await harness.readFile(bashCompletionPath);
          expect(bashContent).toContain('_curl_script_cmd_completion_test');
          expect(bashContent).toContain('complete -F _curl_script_cmd_completion_test');
        });
      });
    });
  }
});

File: tests/e2e/completion_test.go
package e2e

import (
	"testing"
)

func TestE2ECompletion(t *testing.T) {
	t.Parallel()

	ms := NewMockServer(t, "main")
	defer ms.Close()

	h := NewTestHarness(t, HarnessOptions{
		ConfigPath: "config.ts",
		Env: map[string]string{
			"DOTFILES_E2E_USE_REAL_INSTALLERS": "true",
		},
	})
	h.MockServerURL = ms.Server.URL

	// Copy the "main" fixture files to the sandbox TempDir
	h.CopyFixture("main")

	// Run generate command
	stdout, stderr, exitCode, err := h.Generate()
	if err != nil || exitCode != 0 {
		t.Fatalf("generate failed: %v\nstdout: %s\nstderr: %s", err, stdout, stderr)
	}

	// Verify shim was created
	h.AssertShimExistsAndExecutable("curl-script--cmd-completion-test")

	// Install the tool explicitly
	stdout, stderr, exitCode, err = h.Install([]string{"curl-script--cmd-completion-test"})
	if err != nil || exitCode != 0 {
		t.Fatalf("install failed: %v\nstdout: %s\nstderr: %s", err, stdout, stderr)
	}

	// Verify completion files were generated
	zshCompletionPath := ".generated/shell-scripts/zsh/completions/_curl-script--cmd-completion-test"
	bashCompletionPath := ".generated/shell-scripts/bash/completions/curl-script--cmd-completion-test"

	h.AssertFileExists(zshCompletionPath)
	h.AssertFileExists(bashCompletionPath)

	// Verify completion content contains expected strings
	h.AssertFileContentContains(zshCompletionPath, "#compdef curl-script--cmd-completion-test")
	h.AssertFileContentContains(zshCompletionPath, "_curl-script--cmd-completion-test")

	h.AssertFileContentContains(bashCompletionPath, "_curl_script_cmd_completion_test")
	h.AssertFileContentContains(bashCompletionPath, "complete -F _curl_script_cmd_completion_test")
}
